### DISCUSSION TITLE
Moving a pointer up in pointerscope stack

### DISCUSSION THREAD

**Discussion by HGuillemet** [bytedeco/javacpp#724] [Ideas]
Say I have a method doing some complex computation in a `PointerScope` and producing a result as a `Pointer`. We can use `PointerScope.detach(Pointer)` to prevent the result from being deallocated when we exit the scope:
```Java
SomePointer compute() {
  try (PointerScope ps = new PointerScope()) {
    /* Computation, allocating pointers */
    SomePointer result = new SomePointer();
    ps.detach(result);
    /* ... */
    return result;
  }
}
```
But what if the call to `compute` is itself within a `PointerScope` ?
We would like the result to be deallocated when we exit this outper pointer scope.
So what about adding some method `PointerScope.detachUp(Pointer)` to move a pointer from an inner scope to the enclosing scope (if it exists, else simply detach) ? 
Or is there already a way to do it ?

**Comment by saudet** [comment]
There's no way to do that in C++ so it's not something that we should need, but there's a couple of ways we can do something that like that. Either pass the outer scope to compute() and attach it that way before return, or call innerScope.extend() before return, and outerScope.attach() after return.

**Comment by HGuillemet** [comment]
> There's no way to do that in C++ so it's not something that we should need,

Why do you mean by that ? There is no C++ equivalent to `PointerScope`, is it ?
And I'd say the need is driven by the use case. And I believe there is a use case. I'm even close to think that this behaviour could be the default one of `detach`, but changing `detach` could break existing code, so I suggested `detachUp`.

> Either pass the outer scope to compute() and attach it that way before return

That would work, but `compute` may lie in a library and the client code may not know anything about JavaCPP and not depend on it.

>  or call innerScope.extend() before return, and outerScope.attach() after return.

`extend` doesn't take a pointer arg, it will give "ownership" of all pointers allocated in inner scope to the outer scope, right ? 

The workaround I use in such case is to allocate my result object before entering the inner scope, but that's not always possible or easy if I need to do some computations involving pointers to determine arguments needed for allocating the result object.

So nothing essential, but I think it would be a nice addition to JavaCPP that would only add a small number of lines of code.

**Reply by saudet** [reply]
It might be nice for your use case, but what if someone else wants to attach it to outerScope2 instead of outerScope1, or even someCompletelyUnrelatedScope? Your idea doesn't support an important or general enough purpose. If you want to do that, you can use your own extended version of PointerScope tough.

**Comment by HGuillemet** [comment]
The idea is that there might be pointer scopes in the client code, and pointer scopes in the library scope. And the library code shouldn't mess up or assume anything about client code.
The client code must be able to do:
```
SomePointer sp = compute();
```
Just like it does:
```
BytePointer bp = new BytePointer(10);
```
and the resulting pointer should have the same lifespan on both case, depending if the client uses pointer scopes or not.
I find the purpose general enough.

If `compute` uses pointers scopes, we want to detach the returned object from THESE pointer scopes, not to attach to a specific one. But I agree that if `compute` uses nested scopes itself, it could be more general to target a specific scope. Maybe could we directly use `detachUp` from the outer scope even if is pointer is currently attached to the inner one :
```Java
SomePointer compute() {
  SomePointer result;
  try (PointerScope ps1 = new PointerScope()) {
    /* ... */
    try (PointerScope ps2 = new PointerScope()) {
      /* ... */
      result = new SomePointer();
      ps1.detachUp(result);
      /* ... */
    }
    /* ... */
  }
  return result;
}
```
Else we would need to detachUp twice:
```Java
SomePointer compute() {
  SomePointer result;
  try (PointerScope ps1 = new PointerScope()) {
    /* ... */
    try (PointerScope ps2 = new PointerScope()) {
      /* ... */
      result = new SomePointer();
      ps2.detachUp(result);
      /* ... */
    }
    /* ... */
    ps1.detachUp(result);
  }
  return result;
}
```

**Reply by saudet** [reply]
Let's try to figure out something more general

**Comment by HGuillemet** [comment]
Back to this idea.
Struggling with limited GPU memory, and against the non-ref-counting Java GC, I find this feature quite essential.

I have written this function, which doesn't need to patch anything in JavaCPP:

```Java
public static <P extends Pointer> P scopeEscape(P p) {
    if (p == null) return null;
    Iterator<PointerScope> it = PointerScope.getScopeIterator();
    if (it == null) throw new IllegalStateException("Not in a pointer scope");
    PointerScope inner = it.next();
    if (it.hasNext()) {
      it.next().attach(p);
    }
    inner.detach(p);
    return p;
  }
```

And I'm heavily using it like this:
```Java
Tensor forward(Tensor input) {
  try (PointerScope ignored = new PointerScope()) {
     /* Computations creating lots of Pointer subclasses */
    return scopeEscape(result);
  }
}
```

Maybe can you think of something smarter ?

**Reply by saudet** [reply]
The logic is fine, but what something more like this?

```java
Tensor forward(Tensor input) {
  try (PointerScope notIgnored = new PointerScope()) {
     /* Computations creating lots of Pointer subclasses */
    return notIgnored.moveOut(result);
  }
}
```

**Reply by HGuillemet** [reply]
Yes, targetting a specific scope could allow to escape any scope and not only the innermost.

**Reply by saudet** [reply]
Right, but I don't think that's a good thing

**Reply by HGuillemet** [reply]
So maybe a method of `Pointer`: 
```Java
Tensor forward(Tensor input) {
  try (PointerScope ignored = new PointerScope()) {
     /* Computations creating lots of Pointer subclasses */
    return result.escapeScope();
  }
}
```

**Reply by HGuillemet** [reply]
Well, adding any method to `Pointer` is always problematic since it can cause clashes with sublasses.
Why do you think this is not a good thing:
```Java
Tensor forward(Tensor input) {
  try (PointerScope ps1 = new PointerScope()) {
    /* ... */
    try (PointerScope ps2 = new PointerScope()) {
      /* ... */
      return ps1.escape(result);
    }
  }
}
```

**Reply by saudet** [reply]
We can already do that with ps1.attach(result), we don't need anything new for that case.

Actually I think adding a method to Pointer is what makes most sense. It's like "reiniting" the Pointer

**Reply by saudet** [reply]
> ```java
>     return result.escapeScope();
> ```

How about we name it Pointer.exitInnerScope() ?

**Reply by HGuillemet** [reply]
I think "escape" is the coined term for this concept, like in [Escape Analysys](https://en.wikipedia.org/wiki/Escape_analysis). 
`escapeInnerScope` ?

However, there is an issue about my suggested implementation above: if there is no outer scope, then we will `detach` from the inner scope without attaching to another, and since `detach` calls `releaseReference` and the ref counter is 0 afterwards in this case, it will deallocate the pointer. What we need is to remove the pointer from the list of pointers to deallocate when we exit the scope, but not call `releaseReference`. I don't think this is currently possible with the public API.

EDIT: or, rather, decrease the ref counter, since it has been increased when entering the scope, but not call the deallocator if it's 0. BTW I don't understand why `PointerScope` needs to modify the ref counter. Couldl't we simply call the deallocator eagerly at exit if ref counter is null ?

**Reply by saudet** [reply]
Sure, escapeInnerScope() sounds good too.

If there is no outer scope, we can just call retainReference() instead

**Reply by saudet** [reply]
> EDIT: or, rather, decrease the ref counter, since it has been increased when entering the scope, but not call the deallocator if it's 0. BTW I don't understand why PointerScope needs to modify the ref counter. Couldl't we simply call the deallocator eagerly at exit if ref counter is null ?

How do you propose to do reference counting without a reference counter??

**Comment by HGuillemet** [comment]
> We can already do that with ps1.attach(result), we don't need anything new for that case.

`ps.escape(result)` attachs `result` to the scope enclosing `ps` (if any) and detachs `result` from `ps` (or from any enclosing scope). What do you mean ?

**Reply by saudet** [reply]
We can attach a Pointer to multiple scopes, there's no issues with that.

**Reply by HGuillemet** [reply]
Another advantage of binding the method to `PointerScope` and not `Pointer` is that there is no way to escape a scope we don't "own":
Client code:
```Java
try (PointerScope ps = new PointerScope()) {
  // ...
  p = new SomePointer();
  // ...
  compute(p);
  // ...
}
```
Library code:
```Java
void compute(SomePointer p) {
  // ...
  p.escapeInnerScope();
  // ...
}
```
Here the library code messes up with client code by escaping the client scope, while the client code legitimately expects that all the pointers allocated in its scope will be deallocated when scope exits.

With `ps.escape(p)` we must have access to scope `ps` to be able to escape it.

**Reply by saudet** [reply]
Like I keep telling you, we can already to that, and the name of that method is PointerScope.attach(Pointer). What is it that you don't understand?

**Reply by saudet** [reply]
Or are you talking about some method that performs both attach and detach on some random scopes? Like if the user calls someRandomScope.escape(Pointer) on what scope would it call attach()? I don't understand what you're trying to tell me here. You'll need to be more specific

**Reply by HGuillemet** [reply]
So how would we perform this:
```
Tensor forward(Tensor input) {
  try (PointerScope ps1 = new PointerScope()) {
    /* ... */
    try (PointerScope ps2 = new PointerScope()) {
      /* ... */
      return ps1.escape(result);
    }
  }
}
```
with `PointerScope.attach(Pointer)` ?
If we do this:
```
Tensor forward(Tensor input) {
  try (PointerScope ps1 = new PointerScope()) {
    /* ... */
    try (PointerScope ps2 = new PointerScope()) {
      /* ... */
      ps1.attach(result);
      return result;
    }
  }
}
```
Then `result` will survive to the closing of `ps2` but still will be deallocated when `ps1` exits, right ?

**Reply by saudet** [reply]
Yes, so what's your point? I still don't understand how you want to make ps1.escape(result) work.

**Comment by HGuillemet** [comment]
> If there is no outer scope, we can just call retainReference() instead

We don't want the pointer to live forever either. Just to escape the current scope (avoid it for being deallocated when the scope exits).

> How do you propose to do reference counting without a reference counter??

I must be missing something. What will not work if we just register the pointer in the pointer stack of the inner scope when the pointer is created, and call `deallocate` on pointers of the pointer stack when we exit the scope, without doing retain/release reference ?

**Reply by saudet** [reply]
That doesn't allow us to attach the pointer to multiple scopes

**Comment by HGuillemet** [comment]
> What will not work if we just register the pointer in the pointer stack of the inner scope when the pointer is created, and call `deallocate` on pointers of the pointer stack when we exit the scope, without doing retain/release reference ?

I see, attaching to multiple scopes will break. And we are back to your remark about "we can already to that, and the name of that method is PointerScope.attach(Pointer)."  So is ref counting in pointer scopes only useful when the user manually attaches a pointer to some arbitrary scope ?

**Reply by saudet** [reply]
Yeah, that's what reference counting is for

**Comment by HGuillemet** [comment]
> Like if the user calls PointerScope.escape(someRandomScope) on what scope would it call attach()?

Such method would rather be `someRandomScope.escape(somePointer)`) and it would attach the pointer to the scope enclosing `someRandomScope`, like I do in the proposed implementation of `scopeEscape` above, but for an arbitrary scope, not the innermost one.

**Reply by saudet** [reply]
But the Pointer isn't even attached to someRandomScope, I don't see how that could be made to be defined in any non confusing way.

**Reply by saudet** [reply]
Maybe what you want is a someRandomScope.getEnclosingScope(), not sure that's the best name, but for argument's sake, and then we could call attach(somePointer) on it?

**Reply by HGuillemet** [reply]
> But the Pointer isn't even attached to someRandomScope, I don't see how that could be made to be defined in any non confusing way.

Yes, I guess the method should check that the pointer is attached to the scope OR to some enclosed one. But I agree it's not 100% satisfactory.

> Maybe what you want is a someRandomScope.getEnclosingScope(), not sure that's the best name, but for argument's sake, and then we could call attach(somePointer) on it?

Why not, but we can already have the enclosing scope with `PointerScope.getScopeIterator()` and that doesn't solve the issue that, when there is no outer scope, we don't have a way to "deregister" a pointer from current scope without deallocating it.

**Reply by saudet** [reply]
Yeah, so that's why I keep saying, users should pass the scope they want returned objects to be attached to as an argument of their methods.

**Reply by HGuillemet** [reply]
And like I said above, the client code doesn't have to use pointer scope itself, the author can know nothing about JavaCPP and  would like to call:
```
  SomePointer p = SomeLibrary.compute();
```
Just like he calls:
```
  SomePointer p = new SomePointer();
```

It's the author of `compute()` that wants to use pointer scopes and needs a way to return a live pointer.

But if it's not an issue for you and I'm alone to care about such situations... we can stop this discussion. The `scopeEscape` workaround above works enough for me, even if it needs an outer scope, it's more practical than passing a pointer scope to every methods that use pointer scopes and return a pointer.

**Reply by saudet** [reply]
So what about something like this? That seems more reliable, and I don't see how we could smoosh this into a single line anyway.
```java
SomePointer compute() {
    PointerScope returnScope = PointerScope.getInnerScope();
    ...
           if (returnScope != null) returnScope.attach(somePointer);
           return somePointer;
    ...
}
```

**Reply by HGuillemet** [reply]
Where is the pointer scope here ? like this:
```Java
SomePointer compute() {
    PointerScope returnScope = PointerScope.getInnerScope();
    try (PointerScope ps = new PointerScope()) {
        ...
           if (returnScope != null) returnScope.attach(somePointer);
           return somePointer;
        ...
    }
}
```
?
Then this is similar `scopeEscape`:
```Java
public static <P extends Pointer> P scopeEscape(P p) {
    if (p == null) return null;
    Iterator<PointerScope> it = PointerScope.getScopeIterator();
    if (it == null) throw new IllegalStateException("Not in a pointer scope");
    it.next(); // inner scope
    if (it.hasNext()) {
      it.next().attach(p); // outer scope
    }
    return p;
  }
```
Used like this:
```Java
SomePointer compute() {
    try (PointerScope ps = new PointerScope()) {
        ...
           return scopeEscape(somePointer);
        ...
    }
}
```
With the same problem: if `compute` is not called in a pointer scope itself, the method returns a pointer with address 0.

**Reply by saudet** [reply]
We could add a Pointer.resetReference() or something for that

**Reply by HGuillemet** [reply]
We need to remove the pointer from the scope, decrease the reference count, but not call deallocate if the ref count is 0.
What is a bit hard to understand with the reference system (please tell me if I'm wrong), is the following:
If you create pointers outside any scope, they have a ref count of 0 and are perfectly living.
But if you create a pointer in a scope, when the ref count drops to 0, it's automatically deallocated.
So there are 2 kinds of pointers "in-scope" and "out-of-scope".

I guess your `resetReference` suggestion is meant to turn an "in-scope" pointer into an "out-of-scope", living, 0-count, pointer.
But you also need to know the scopes to which it has been attached in order to remove it from their `pointerStack`. How would you do that ?
Or, we establish a rule to distinguish "in-scope" and "out-of-scope" pointers: a pointer is out-of-scope iif its count is -1, or
a pointer is ouf-of-scope iif it's living and its count is 0. And in `DeallocatorReference.release`, when you recognize an out-of-scope pointer, you do nothing.
In this case, after a `resetReference` or whatever it's called, you can leave the pointers attached to some scopes and nothing will happen when the scopes exit.
But what if you attach again to a scope ? That's not really satisfactory.

I think it would be better to add another version of `detach` that doesn't deallocate if the count reaches 0.  For instance an overload with a boolean argument. We would need to also introduce a `Pointer.releaseReference` with a boolean argument and a `DeallocatorReference.release` with a boolean argument.

EDIT: in fact this `PointerScope.detach(Pointer, boolean)` could rather be `PointerScope.escape(Pointer)`, but we still need a `Pointer.releaseReference(boolean)` and `DeallocatorReference.release(boolean)`.

**Reply by saudet** [reply]
Your understanding of the relationship between the reference counter and pointer scopes is accurate, so I don't really see a nice way to make everything that you want it do work. We can add some other version of detach(), sure, but we still need to know which scope it has been attached to. If the user doesn't know, then I don't see a clean solution.

**Reply by HGuillemet** [reply]
The user knows the scope he wants to detach from. 

To sum it up, here is my suggestion:
* Create `Pointer.releaseReference(boolean deallocate)` and `DeallocatorReference.release(boolean deallocate)` that do the same thing that current versions of these methods, but trigger deallocation only if ref count drops to 0 AND `deallocate` is true. Add no-arg shortcuts for backward compatibility.
* Then 3 options:
1) Add an instance method `PointerScope.escape(Pointer)` that removes the pointer from the `pointerStack` of this scope, calls `releaseReference(false)` on the pointer, and attaches the pointer to the scope enclosing this scope, in the current thread, IF ANY. If the pointer is not in the `pointerStack` of this scope, throws an exception.
2) Add an instance method `PointerScope.escape(Pointer)` that does the same thing, but this scope can be any scope of the current thread. The pointer is removed from the `pointerStack` of any scope from the current innermost scope of this thread up to this scope. This allows to escape from 2 scopes with a single call for convenience. Throws an exception if the pointer is not found in any of these `pointerStack`.
3) Add a static method `escapeScope(Pointer)` that does the same thing than 1) but on `getInnerScope()`.

My preference is for 2).

**Reply by saudet** [reply]
What if instead we could set Pointer objects into some "return mode" by calling returnReference() or something like that in which when the counter reaches 0, instead of deallocating, it would first try to attach itself to the new inner scope, or failing that, remain in an alive state at 0? Then we could achieve what you want by calling something like `somePointerScope.attach(somePointer.returnReference())` or through some special version of attach() like `somePointerScope.attachAndReturn(somePointer)` or maybe the other way around `somePointer.returnReference(somePointerScope)`? Doesn't feel exactly right, but still user-friendler than than a new concept of "escape".

**Reply by HGuillemet** [reply]
What is the role of `somePointerScope` in your proposal ? How would that be used in the typical use case ?

**Reply by HGuillemet** [reply]
If `somePointerScope` is the scope we want to "escape" from then this `attach` looks pretty well like the `PointerScope.escape` I'm suggesting, but why calling this attach ? The pointer has already been attached automatically to the scope when created.
